### ISSUE TITLE
Question about custom descriptors

### ISSUE THREAD

**Issue by eddienewton** [colmap/colmap#772]
Hi,

Great work on Colmap!  I had a question about importing custom features.  

The tutorial states descriptor values should be between 0 and 255.  I'm using a custom descriptor with a -1 to 1 value range.  

There is a note here that states to do a L2 normalization of the descriptor:
https://github.com/tsattler/visuallocalizationbenchmark/tree/master/local_feature_evaluation

Is that basically a L2 normalization of each descriptor row?  Or is the normalization scope all descriptors of all images, or all descriptors to one particular image?

Thanks in advance.

**Comment by eddienewton** [comment]
From what I'm gathering, I need to do the following:

- Ensure each descriptor row sums up to 1.0.
- Find the maximum and minimum descriptor values for the entire image set, and use those to normalize descriptor values between 0 and 255.

Finally, I'm assuming its ok if the descriptors are float values between 0 and 255?

Thanks

**Comment by eddienewton** [comment]
I went ahead and tested importing my custom features.  I did the following to convert my descriptors to Colmap:

- L2 descriptor row normalization.  Each row has a norm of 1.0.
- Rescaled the descriptors to 0 to 255 using global project min and max descriptor values.
- Converted float descriptors to uint8.

Unfortunately, Colmap isn't finding any matches.

During troubleshooting using OpenCV, I did notice that brute force using NORM_L2 and the original float descriptors produces great matching.  However, when I use brute force using NORM_HAMMING, and the normalized uint8 descriptors, my matching degrades tremendously.  This may be causing Colmap to find no matches.

Any thoughts or suggestions?

Thanks.

**Comment by eddienewton** [comment]
Probably can close this issue.  It turns out since my descriptor is a zero-mean, I wasn't able to easily convert the descriptor to the 0-255 scaling, while maintaining the same distance measurements.  In the end, I was able to get some matching results from my custom descriptor, but the results were different than if I just used OpenCV Flann or brute force matching.  

I'll close this issue, and will import my matches into Colmap.
